### DISCUSSION TITLE
How to extend the ChatHistory class?

### DISCUSSION THREAD

**Discussion by MalayAgr** [microsoft/semantic-kernel#9931] [Q&A] [answered]
I have a scenario where I want to make the ChatHistory class an observable so that other classes can subscribe to changes in the ChatHistory. The extended class would notify the subscribers whenever a message is added to the ChatHistory.

Unfortunately, I just realised that none of the methods used to add messages to the ChatHistory are virtual and hence, cannot be overridden. I was hoping to leverage polymorphism so that the rest of my code doesn't need to change and I can use this observable ChatHistory and the usual ChatHistory class interchangeably.

How can I achieve this?

**Comment by sphenry** [comment, marked as answer]
@MalayAgr thanks for this, it's some we should think about supporting in the framework. I'm converting this into a backlog issue

**Reply by MalayAgr** [reply]
Hi, thanks for taking this up! Should I assume that this is not possible for currently?

**Reply by dmytrostruk** [reply]
@MalayAgr It depends on your scenario. If you are interested in publisher/subscriber solution outside of `IChatCompletionService`, instead of `ChatHistory` you can use `IList<ChatMessageContent>`. If the scenario is related to `IChatCompletionService`, then I'm not sure that `ChatHistory` can be extented with this functionality as for now. 

But there are other ways, like implementing new `IChatCompletionService` or creating a decorator for it, so it can produce notifications when needed. Filters can also help if your case includes function calling. 

If you could share your scenario in more detail, we can try to find some alternative solution. Otherwise, we will think about making `ChatHistory` methods more extendable. Thank you!
